### PR TITLE
Fatal error if no containers for service or task

### PR DESCRIPTION
**Pull request by ken9ross**
## Intent of Change

- New feature (non-breaking change which adds functionality)

## Description
Raise Hamlet fatal error for task without container 

## Motivation and Context
Fixes https://github.com/hamlet-io/engine-plugin-aws/issues/361
Shift onus of test onto Hamlet

## How Has This Been Tested?
Locally

### Prerequisite PRs:

- None

### Dependent PRs:

- None

### Consumer Actions:

- None

### CODE REVIEW

File: aws/components/ecs/setup.ftl [hunk 1]
```diff
@@ -1611,17 +1611,21 @@
...
+                    engine=engine
+                    containers=containers
+                    role=roleId
+                    executionRole=executionRoleId
+                    networkMode=networkMode
+                    dependencies=dependencies
+                    fixedName=solution.FixedName
+                /]
+            [#else]
+                [@fatal message="No container to deploy services or tasks into" /]
```
Comment (roleyfoley) [inline aws/components/ecs/setup.ftl:1627]: The exception probably needs some more context.

- The containers belong to the task/service not the other way around  
- The message should be a quick message outlining what the problem was 
- The context of the message should include things like taskId 

You also don't need to make creating the task conditional. The createTemplate process will fail as a fatal exception has been added to the log messages. Making the task creation conditional means that we can't look at the fatal template during debugging. 

so might be better to just put the if conditional around the fatal message

**Review by roleyfoley — CHANGES_REQUESTED** [review]

File: aws/components/ecs/setup.ftl (hunk 1 above)
Comment (ken9ross) [reply aws/components/ecs/setup.ftl:1627]: Cool

File: aws/components/ecs/setup.ftl [hunk 2]
```diff
@@ -1622,6 +1622,10 @@
                 dependencies=dependencies
                 fixedName=solution.FixedName
             /]
+            
+            [#if containers?size < 1 ]
+                [@fatal message="No container available" context=resources["task"] /]
```
Comment (roleyfoley) [inline aws/components/ecs/setup.ftl:1627]: Is that context useful enough to be able to identify the task or service where this has happened? 
If you looked at what is provided could you fix the issue?

Comment (ken9ross) [reply aws/components/ecs/setup.ftl:1627]: That sounds like I'm missing something obvious. I get an error message of
```
{
  "Timestamp": "2021-08-02T08:31:24.602+10:00",
  "Severity": "fatal",
  "Message": "No container available",
  "Context": {
    "Id": "ecsTaskXappXserviceecsbaseXtaskbase",
    "Name": "application-serviceecsbase-taskbase",
    "Type": "ecsTask",
    "Deployed": false
  }
}
```
With a solution file of
```
  "Tiers": {
    "app": {
      "Components": {
        "serviceecsbase": {
          "ecs": {
<snip>
            "Tasks": {
              "taskbase": {
                "Instances": {
                  "default": {
                    "deployment:Unit": "aws-task-base"
                  }
                },
<snip>
```

Comment (roleyfoley) [reply aws/components/ecs/setup.ftl:1627]: From the exception message could I figure out what I need to do to fix the problem that has caused a fatal error to occur? 

I think it needs a suggestion on what to do or a hint on whats going on. Its also containers not container. You can have multiple containers in a task or service definition

Comment (ken9ross) [reply aws/components/ecs/setup.ftl:1627]: Like
```
{
  "Timestamp": "2021-08-02T16:49:20.496+10:00",
  "Severity": "fatal",
  "Message": "No container available. Add one or more containers to the following service/task",
  "Context": {
    "Id": "ecsTaskXappXserviceecsbaseXservicebase",
    "Name": "application-serviceecsbase-servicebase",
    "Type": "ecsTask",
    "Deployed": false
  }
}
```
Or am I still missing the point

Comment (roleyfoley) [reply aws/components/ecs/setup.ftl]: Yeah I think that covers it.

**Review by roleyfoley — APPROVED** [review]

**Pull request merged**